### ISSUE TITLE
[FEAT]: Add support for GitHub Advanced Security features

### ISSUE THREAD

**Issue by samtarplee** [integrations/terraform-provider-github#2043]
### Describe the need

Github Advanced Security provides a lot of features, such as secret scanning, code scanning, and more expansive Dependabot management.
Being able to manage these features for a GitHub organisation at scale using Terraform would be very useful, particularly the secret scanning, which can support custom patterns.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by nnellanspdl** [comment]
What specifically do you need?  I currently use the following GitHub Advanced Security features in my code:

```terraform
resource "github_repository" "symbolic_name" {
  
  vulnerability_alerts = true

  security_and_analysis {
    advanced_security {
      status = "enabled"
    }
    secret_scanning {
      status = "enabled"
    }
    secret_scanning_push_protection {
      status = "enabled"
    }
  }
}

resource "github_repository_dependabot_security_updates" "symbolic_name" {
  enabled = true
}

**Comment by samtarplee** [comment]
@nnellanspdl Being able to manage custom patterns for secret scanning is the main one I'd like, as well as managing the above options on an organisation level.

**Comment by IMax153** [comment]
I agree - having the ability to configure these options on an organizational level would be extremely useful for the [Effect-TS](https://github.com/Effect-TS) GitHub organization.

**Comment by kfcampbell** [comment]
Is this supported in the [API](https://docs.github.com/en/rest/secret-scanning/secret-scanning?apiVersion=2022-11-28)? I'm not seeing it.

**Comment by ChristopherMank** [comment]
Would love to also see the ability to manage:
1. Scan for non-provider patterns
2. Use AI detection to find additional secrets
3. Validity checks
4. Autofix for CodeQL

**Comment by ChristopherMank** [comment]
Not stale, still very much want to see this.

**Comment by slw07g** [comment]
> Is this supported in the [API](https://docs.github.com/en/rest/secret-scanning/secret-scanning?apiVersion=2022-11-28)? I'm not seeing it.

Yes, it's supported in the API
https://docs.github.com/en/rest/code-scanning/code-scanning?apiVersion=2022-11-28#update-a-code-scanning-default-setup-configuration

**Comment by anGie44** [comment]
At the enterprise and org level, looks like we can leverage https://docs.github.com/en/rest/code-security/configurations?apiVersion=2022-11-28

**Comment by deiga** [comment]
@slw07g @anGie44 We'd appreciate PRs for these! :)

**Comment by priiiiit** [comment]
This actually is now more relevant, since `/enterprises/{enterprise}/code_security_and_analysis` [has been removed](https://docs.github.com/en/enterprise-cloud@latest/rest/enterprise-admin/code-security-and-analysis?apiVersion=2022-11-28)(?), but current version of the provider is still trying to use that and failing, making resource `github_enterprise_security_analysis_settings` essentially broken.

**Comment by fishpen0** [comment]
The current support matrix I am working around with hooks to scripts

requirement | Provider support | Terraform attribute
-- | -- | --
Secret protection | Yes | secret_scanning { status = "enabled" }
Validity checks | No | NA
Extended metadata | No | NA
Non-provider patterns | Yes | secret_scanning_non_provider_patterns { status = "enabled" }
Scan for generic passwords | Yes | secret_scanning_ai_detection { status = "enabled" }
Push protection | Yes | secret_scanning_push_protection { status = "enabled" }
Bypass Groups | No | NA
Prevent direct dismissals | No | NA

**Comment by deiga** [comment]
@priiiiit could you open a new issue since it's distinct from this issue?

**Comment by NicolasAizier** [comment]
It has been 3 years now and I can confirm that the original answer from remain true.
The `security_and_analysis` block is supposed to have 6 sub-block to enabled 6 features.
1. There are far more than 6 blocks needed to covered all the options that would be useful for a real Organization as code control
2. only 3 works, the 3 others return and systematic update wth no consequence

Those are the working ones
```
resource "github_repository" "some_repo" {
  
  security_and_analysis {
    advanced_security {
      status = "enabled"
    }
    secret_scanning {
      status = "enabled"
    }
    secret_scanning_push_protection {
      status = "enabled"
    }
  }
}
```

**Comment by vbalumuri** [comment]
Bumping feature requests for visibility and traction. This is a great feature to have at the organization level, allowing you to manage the default security and analysis config and attach repositories or blanket-apply it to all repositories within the organization, rather than configuring each repository individually.
